File: telegram-bot/src/handlers/message.py
import re

from .. import currency, db, pending_rate
from ..convert import send_conversion
from ..format import code

AMOUNT_CODE_RE = re.compile(r'^(-?\d+(?:\.\d+)?)\s+([A-Za-z]{2,10})$')
BARE_CODE_RE = re.compile(r'^([A-Za-z]{2,10})$')


async def message_handler(event):
    text = (event.raw_text or '').strip()
    db.ensure_user(event.sender_id)

    amount_match = AMOUNT_CODE_RE.match(text)
    if amount_match:
        pending_rate.consume(event.sender_id)
        return await _handle_convert(event, float(amount_match.group(1)), amount_match.group(2))

    bare_match = BARE_CODE_RE.match(text)
    if bare_match and pending_rate.consume(event.sender_id):
        return await _handle_convert(event, 1, bare_match.group(1))

    # Not a recognised format, and not something we're expecting - stay quiet
    # rather than nagging on every unrelated message in a group chat.


async def _handle_convert(event, amount, raw_code):
    currency_code = raw_code.lower()
    valid = await currency.is_valid_currency(currency_code)
    if not valid:
        return await event.respond(
            f"I don't recognise the currency {code(currency_code.upper())}. Check the code and try again.",
            parse_mode='md',
        )
    await send_conversion(event, event.sender_id, amount, currency_code)

File: telegram-bot/src/handlers/message.js
const db = require('../db');
const currency = require('../currency');
const pendingRate = require('../pendingRate');
const { sendConversion } = require('../convert');
const { code } = require('../format');

const AMOUNT_CODE_RE = /^(-?\d+(?:\.\d+)?)\s+([A-Za-z]{2,10})$/;
const BARE_CODE_RE = /^([A-Za-z]{2,10})$/;

async function messageHandler(ctx) {
  const text = (ctx.message.text || '').trim();
  db.ensureUser(ctx.from.id);

  const amountMatch = text.match(AMOUNT_CODE_RE);
  if (amountMatch) {
    pendingRate.consume(ctx.from.id);
    return handleConvert(ctx, parseFloat(amountMatch[1]), amountMatch[2]);
  }

  const bareMatch = text.match(BARE_CODE_RE);
  if (bareMatch && pendingRate.consume(ctx.from.id)) {
    return handleConvert(ctx, 1, bareMatch[1]);
  }

  // Not a recognised format, and not something we're expecting - stay quiet
  // rather than nagging on every unrelated message in a group chat.
}

async function handleConvert(ctx, amount, rawCode) {
  const currencyCode = rawCode.toLowerCase();
  const valid = await currency.isValidCurrency(currencyCode);
  if (!valid) {
    return ctx.replyWithMarkdownV2(`I don't recognise the currency ${code(currencyCode.toUpperCase())}\\. Check the code and try again\\.`);
  }
  await sendConversion(ctx, ctx.from.id, amount, currencyCode);
}

module.exports = messageHandler;
